tests/EventProjectorServiceProviderTest.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace Tests\Spatie\EventProjector;
+
+use ReflectionClass;
+use Spatie\EventProjector\Tests\TestCase;
+use Spatie\EventProjector\EventProjectionist;
+use Spatie\EventProjector\Projectors\Projector;
+use Spatie\EventProjector\Projectors\ProjectsEvents;
+
+class EventProjectorServiceProviderTest extends TestCase
+{
+    /** @var \Spatie\EventProjector\EventProjectionist */
+    private $eventProjectionist;
+
+    public function setUp()
+    {
+        parent::setUp();
+
+        $this->eventProjectionist = app(EventProjectionist::class);
+    }
+
+    /** @test */
+    public function projectors_can_be_registered_via_config()
+    {
+        /** @var \Illuminate\Support\Collection $projectors */
+        $projectors = $this->getProtectedProperty($this->eventProjectionist, 'projectors');
+
+        $this->assertTrue($projectors->isNotEmpty());
+
+        $this->assertEquals(DummyProjector::class, $projectors->first());
+    }
+
+    /** @test */
+    public function reactors_can_be_registered_via_config()
+    {
+        /** @var \Illuminate\Support\Collection $reactors */
+        $reactors = $this->getProtectedProperty($this->eventProjectionist, 'reactors');
+
+        $this->assertTrue($reactors->isNotEmpty());
+
+        $this->assertEquals(DummyReactor::class, $reactors->first());
+    }
+
+    protected function getProtectedProperty($class, $property)
+    {
+        $reflection = new ReflectionClass($class);
+
+        $property = $reflection->getProperty($property);
+
+        $property->setAccessible(true);
+
+        $propertyValue = $property->getValue($class);
+
+        return $propertyValue;
+    }
+
+    protected function getPackageProviders($app)
+    {
+        $app['config']->set('event-projector.projectors', [
```

Comment:
Let's just set the config to the right value in the `setUp`method of this class.

Reply:
The correct place in Orchestra would be `getEnvironmentSetUp`

Reply:
The problem here is that it seems that both `setUp` and `getEnvironmentSetUp` are run after the service providers are registered, and thus the tests fail. Any other suggestions?

Reply:
You might be able to fix this by using `resolving` instead of the closure in the  `singleton` call in `EventProjectorServiceProvider`.

```php
$this->app->resolving(HelpSpot\API::class, function ($api, $app) {
    // Called when container resolves objects of type "HelpSpot\API"...
});
```

Reply:
Is it worth changing the service provider for this though? I'd rather have a test doing some workarounds than the package code itself.

Reply:
Either way: changing config in tests van be a pain in the ass, been bitten by this a few times in the past...

Reply:
Tested this, but it also doesn't work with `resolving`; it's ran before `getEnvironmentSetUp` 🤔

Reply:
Did some more debugging in Orchestra, as far as I can tell this is the only place where I can inject configuration before the service providers are loaded.

```php
    protected function resolveApplicationConfiguration($app)
    {
        $app->make('Illuminate\Foundation\Bootstrap\LoadConfiguration')->bootstrap($app);

        tap($this->getApplicationTimezone($app), function ($timezone) {
            ! is_null($timezone) && date_default_timezone_set($timezone);
        });

        $app['config']['app.aliases'] = $this->resolveApplicationAliases($app);
        $app['config']['app.providers'] = $this->resolveApplicationProviders($app);
    }
```

This is an excerpt from `\Orchestra\Testbench\Concerns\CreatesApplication`, the config is only loaded at the start of this function, so we cannot add it before that line. However: the service providers are loaded at the end of this function, so there's only two points where we can hook into: `resolveApplicationAliases` and `resolveApplicationProviders`; I feel that `resolveApplicationProviders` is the only right option..

Reply:
Think it's worth PR'ing a `getApplicationConfiguration` hook to Orchestra?

Reply:
I've sent a PR: https://github.com/orchestral/testbench-core/pull/12

Reply:
There few thing that I see doesn't feel right here.

```php

        $this->app
            ->when(EventSubscriber::class)
            ->needs('$storedEventModelClass')
            ->give(config('event-projector.stored_event_model'));
        $this->app
            ->when(ReplayEventsCommand::class)
            ->needs('$storedEventModelClass')
            ->give(config('event-projector.stored_event_model'));
```

And 

```php
$this->mergeConfigFrom(__DIR__.'/../config/event-projector.php', 'event-projector');
```

Why isn't this being done in `boot()` instead of `register()`. This cause `config('event-projector')` to be resolved much earlier in the services registration flow. Ideally all service providers should just register services in `register()` and handle bootstrapping it under `boot()`.